### ISSUE TITLE
What next?

### ISSUE THREAD

**Issue by Megalomaniak** [TwistedTwigleg/vanilla-bootstrap#17]
I mean after we have called this current theme release "finished"?

Probably should rewrite a whole new theme from scratch anyways since bootstrap is built using some outdated concepts that are deprecated with replacements existing, or we could just try to refactor bootstrap fully, but that is probably a more difficult task.

_**either ways this should be a future consideration**_

**Comment by Megalomaniak** [comment]
Basically I see it this way:

We should just tone down this theme/release target, get it done A.S.A.P. while K.I.S.S. and then instead of slowly working away at next version release of this I'd rather start work on a new theme from scratch. This bootstrap theme was developed before themes and plugins were merged into a common concept(addons). 

Bootstrap still does things in old and essentially deprecated ways. Not to mention that the current bootstrap is itself a rewrite of an older bootstrap based theme, moving from bootstrap (I guess) 2.x to bootstrap 3. Now bootstrap 3 is deprecated and bootstrap 4 is out.

We could use bootstrap 4 as a basis and start working on a new godot specific theme that is referencing what remains relevant from the old bootstrap theme.

**Comment by TwistedTwigleg** [comment]
I agree that we should try to get the theme done ASAP. Ideally I'd like to migrate the forums over in a week or two. Right now the test forums is more or less stable, albeit it has not really been stress tested. 

We need to figure out what needs to be done to fully migrate the forums, but that is a discussion we should include the others on. I'll message everyone about it soon, once I've gathered my thoughts and have the time to write down a message.

As for making a new theme, I'm not 100% sure. Here are my thoughts on the matter:
_____

Making a new theme will take a lot of time, especially if we do it from scratch. Don't get me wrong, the customization with making a new theme is tremendous and would allow us to do all sorts of cool things, but it also means that the heavy work would be on us to get it done.

One huge plus is we can make the theme *exactly* like we want. We should in theory be able to make the forums look as we envision, with the only limitation being our time, energy, and knowledge of CSS, PHP, and Javascript.

That said, this is also one of the biggest downsides too. We'd have to make everything ourselves, from the littlest pieces to the most complex. It is probably going to take quite a bit of setup work before we'll be able to see nice looking results.

This also means we'd be changing everything up for the forum users again, which may not be a bad thing but is something to consider. We'd also have to write support for different screen sizes and for mobile/touch devices ourselves, for better or for worse.

To be fair, I haven't actually made a custom Vanilla theme myself, so maybe it is not that hard and wouldn't take long. I really don't know, but I imagine it will take longer than editing the Bootstrap theme has taken.

______

We could probably alleviate some of the work by editing a more modern theme, like Keystone, instead. This would certainly speed things up development wise and would give us a nice setup to fall back on.

Editing a modern theme  would still give us flexibility and control while allowing us to get started quickly and not have to reinvent the wheel entirely. Especially if we choose a theme that doesn't use a framework, unlike the Bootstrap theme, then we have just as much control as if we wrote one from scratch as far as customization goes.

_______

Regardless, we'd basically be starting from scratch again, which is not necessarily a bad thing or something I'm per say against. However, this leads me to this question: **Would a new theme offer a great enough benefit that it is worth the added time and energy creating a new theme would take, or is it better to just customize what we have and move on.**

Don't get me wrong, I certainly agree that making/editing a new theme would be cool, would offer advantages, and would allow us to really make the theme something special. But on the flip side, it will also take lots of time, effort, and energy to get it to that point.

That is not a bad thing, nor am I against it, but I'm not sure how much time and energy I'll be able to dedicate in the future, especially with things like work and collage thrown in.

That said, if we want to make/edit a new theme then I'll totally help, I'm by no means against the idea or going to stop helping, I'm just not sure how much time/energy I'll be able to pour into it in the near-ish future and I don't want to say "Let's do it", make plans, and then be unable to help and have to leave the majority of the work on you.

All of that above said, I'm not necessarily saying let's stick with Bootstrap either. The only reason I have suggested using Bootstrap is because it was what the forums were using before, and it made things like t heme changing possible in a quick and efficient manner. I think it works well enough for our purposes, and while it is clunky and dated, it is still for the most part workable.

That's just my opinion on the matter. I've got to go, so I haven't really edited everything very much. **I'm not against the idea by any means, I just don't know if I'll be able to help, nor if the benefits are worth the extra time and effort it would take.** (not to mention we could be putting that time/energy into other things/projects)

**Comment by Megalomaniak** [comment]
> Making a new theme will take a lot of time, especially if we do it from scratch. Don't get me wrong, the customization with making a new theme is tremendous and would allow us to do all sorts of cool things, but it also means that the heavy work would be on us to get it done.

I would expect for us to have however long we would need to finish it, since the theme would only go into use once done. **And the whole point is for the current one we are working on to be again a stable one we can use for a year or two.** But bootstrap will be increasingly hard to maintain as vanilla has updates over time.

During the time we are using the current, we can start low-key working on the next one.

> That said, this is also one of the biggest downsides too. We'd have to make everything ourselves, from the littlest pieces to the most complex. It is probably going to take quite a bit of setup work before we'll be able to see nice looking results.

Yes, and no. Everything in the theme addon, yes. Beyond that external addons, no. More importantly any plugins we currently use that aren't any longer maintained we will/already have to maintain ourselves anyways.

So choosing which of them we deem important enough to keep the functionality of we can and perhaps should then instead integrate directly into the new addon and document for maintainability(_for should the forum/project change hands again in the future..._).

> This also means we'd be changing everything up for the forum users again, which may not be a bad thing but is something to consider. We'd also have to write support for different screen sizes and for mobile/touch devices ourselves, for better or for worse.

from my previous message:
> We could use bootstrap 4 as a basis and start working on a new godot specific theme that is referencing what remains relevant from the old bootstrap theme.

I expect that to be quite a bit still.

---

> We could probably alleviate some of the work by editing a more modern theme, like Keystone, instead. This would certainly speed things up development wise and would give us a nice setup to fall back on.

Yes, looking at keystone as another additional reference is not a bad idea either since it should be already coded with migration to the future/new vanilla template-engine[<sup>1</sup>] in mind. _And unlike current bootstrap it is maintained and will continue to get updated for the foreseeable future._

---

> Regardless, we'd basically be starting from scratch again, which is not necessarily a bad thing or something I'm per say against.

With vanilla this is actually easier than you might initially expect, especially in cases where we have already one or more themes to look at for reference.

> However, this leads me to this question: **Would a new theme offer a great enough benefit that it is worth the added time and energy creating a new theme would take, or is it better to just customize what we have and move on.**

Customizing code developed by someone else is honestly more difficult and harder to maintain than just writing it yourself in the first place. In this case we do have the git history to look at as sort of "documentation" however it's certainly not optimal.

But again, as I said above at the start, we would likely be sticking with the current thing for quite a while anyways. This discussion is really about the longer term outlook, think 2 or 3 years from now as vanilla forums 3.0 will likely be out with possibly quite a few breaking changes in it.

[<sup>1</sup>]: https://open.vanillaforums.com/discussion/comment/251706/#Comment_251706

**Comment by TwistedTwigleg** [comment]
> I would expect for us to have however long we would need to finish it, since the theme would only go into use once done. **And the whole point is for the current one we are working on to be again a stable one we can use for a year or two.** But bootstrap will be increasingly hard to maintain as vanilla has updates over time.
> During the time we are using the current, we can start low-key working on the next one.

Ah, I didn't realize it would be that long term, I was thinking more in the months range.
In that case, then I do not have any problems with making a new theme to eventually replace the current one.

**Comment by Megalomaniak** [comment]
Considering how slowly vanilla is progressing right now, what with everything going on in the world I think it's safe to assume we wont be facing any breaking changes to the theming any time soon so iterating on the current is probably fine. Only consider building a new theme from scratch then once we know that a new theming and templating engine is coming and more importantly when exactly it would be released.

**Comment by TwistedTwigleg** [comment]
Sounds good. Right now the theme we are using seems to be working okay. Though, I do like the idea of remaking the theme from scratch if the one we are using becomes incompatible in the future.
